core/block/editor/state/objecttypes.go

```diff
@@ -0,0 +1,191 @@
+package state
+
+import (
+	"strings"
+
+	"github.com/anyproto/anytype-heart/core/block/undo"
+	"github.com/anyproto/anytype-heart/core/domain"
+	"github.com/anyproto/anytype-heart/pb"
+	"github.com/anyproto/anytype-heart/pkg/lib/localstore/addr"
+	"github.com/anyproto/anytype-heart/util/slice"
+)
+
+type ChangeApplier interface {
+	ApplyChanges(ch *pb.ChangeContent)
+}
+
+type ChangeGetter interface {
+	Diff() (ch []*pb.ChangeContent)
+}
+
+type ObjectType interface {
+	ChangeApplier
+	ChangeGetter
+	ObjectTypeKeys() []domain.TypeKey
+	ObjectTypeKey() domain.TypeKey
+	SetObjectTypeKeys(objectTypeKeys []domain.TypeKey)
+	SetParent(parent ObjectType)
+	SetObjectTypeKey(objectTypeKey domain.TypeKey)
+	SetNoObjectType(noObjectType bool)
+	GetObjectTypeHistory() *undo.ObjectType
+	SetParentObjectType()
+}
+
+type ObjectTypes struct {
+	objectTypeKeys []domain.TypeKey
+	parent         ObjectType
+	noObjectType   bool
```

Comment:
Is noObjectType necessary here?

Reply:
Yes, it also related to object type component in state. It also used in `ObjectTypeKey()` function